// node-jt400/dist/lib/ifs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ifs = ifs;
const path_1 = require("path");
const read_stream_1 = require("./read_stream");
const write_stream_1 = require("./write_stream");
function ifs(connection, bufferToJavaType, javaTypeToBuffer) {
    return {
        createReadStream: function (fileName) {
            const javaStream = Promise.resolve(fileName).then(function (file) {
                return connection.createIfsReadStream(file);
            });
            return new read_stream_1.IfsReadStream({
                ifsReadStream: javaStream,
                javaTypeToBuffer,
            });
        },
        createWriteStream: function (fileName, options = { append: false }) {
            const javaStream = Promise.resolve(fileName).then(function (file) {
                const folderPath = (0, path_1.dirname)(file);
                const fileName = (0, path_1.basename)(file);
                return connection.createIfsWriteStream(folderPath, fileName, options.append, options.ccsid);
            });
            return new write_stream_1.IfsWriteStream({
                ifsWriteStream: javaStream,
                bufferToJavaType,
            });
        },
        deleteFile: (fileName) => connection.deleteIfsFile(fileName),
        fileMetadata: (fileName) => connection.getIfsFileMetadata(fileName).then(JSON.parse),
    };
}
